### PR TITLE
Rewritten part of File Cache Storage Engine

### PR DESCRIPTION
**Pull request by demis-palma**
### Definitive solution for “zero length cache file”
that has caused so much troubles in such a short time (and so many problems would have caused in future) https://github.com/joomla/joomla-cms/pull/10767 https://github.com/joomla/joomla-cms/issues/15544 https://github.com/joomla/joomla-cms/pull/15558 https://github.com/joomla/joomla-cms/pull/15592

The current implementation of file storage makes use of the shutdown function, but it doesn't work in case of software crashes, hardware failure, power blackout, and so on.

## Summary of Changes

1. Don't create empty cache files at all
2. Limited the lock for no longer than is strictly necessary to the write operation
3. Delegated file locking to PHP
4. Addressed and fixed a couple of other latent cache bugs
5. Significant code simplification and major performance improvements.
6. Given resistance and stability against the unlikely event of corrupted cache files (not only zero length)

#### 1. Write followed by a concurrent Write
:anguished: **Before this PR** It depends on how long it takes the first write process. If the first write ends before the second’s time-out, then the concurrent write will be enqueued, otherwise it will be discarded. A big mess.
:heavy_check_mark: **After this PR** The concurrent write will be enqueued to the first write in any case.

#### 2. Write followed by a concurrent Read
:x: **Before this PR** Unserialize error. This also causes a PHP warning, which could possibly disclosure path information.
This happens because the caller expects a valid serialized data or null. It does not expect invalid data.
Unlikely to actually happen? For the “[Infinite monkey theorem](https://en.wikipedia.org/wiki/Infinite_monkey_theorem)” it will almost surely happen and actually happened in past: https://forum.joomla.org/viewtopic.php?t=851183
:heavy_check_mark: **After this PR** No error happens. The cache ensures to always return valid data or null, as the caller expect. This is done by introducing an integrity self check to the cache buffer.

#### 3. Read followed by a concurrent Read
:heavy_check_mark: **Before this PR** No problem
:heavy_check_mark: **After this PR** No problem

#### 4. Read followed by a concurrent Write
:exclamation: **Before this PR** It depends on the file size. The most of the times due to the hard drive buffer, the entire file remains available in the drive cache for the reading process even if it has been already overwritten on the disk.
Otherwise, if the file is bigger than the hard drive cache, the cache buffer gone corrupted and it causes unserialize error to the next process that will read the file (see above).
:heavy_check_mark: **After this PR** The same as before the PR, with the difference that a corrupted file is detected and invalidated by the integrity self check, and it raises no errors. The normal program flux will continue, ignoring the cache for the current http request.

#### 5. Write followed by a concurrent Deletion
Quite scary, but no particular problem. It is behaves as case 1. Write followed by a concurrent Write

#### 6. Read followed by a concurrent Deletion
Quite scary, but no particular problem. It is behaves as case 4. Read followed by a concurrent Write

#### 7-8. Deletion followed by a concurrent Read | Write
Deletion is atomic and does not suffer concurrency from Read or Write operations, so the cases Deletion followed by a concurrent Read | Write don’t exist.

## Performance improvements

1. Do not delete anymore expired files before recreate them. Just overwrite them.
2. Do not use anymore the str_replace() function searching for the php die protection heading along the entire cache buffer. Used an offset to shift the file pointer + sequential reading instead.

## Other changes
Mimic a successful lock and unlock operations to keep compatibility with the previous interfaces, but this storage engine does not need lock / unlock methods

## Testing Instructions
I have stressed my system with 100 concurrent requests a a time, and I got no errors.
At the moment, I've tested on Linux only. Windows and OSX should not reserve surprises, but It's better to do a test drive.

### CODE REVIEW

File: libraries/joomla/cache/storage/file.php [hunk 1]
```diff
@@ -26,13 +26,32 @@ class JCacheStorageFile extends JCacheStorage
...
+	 * to reflect the actual size of this constant.
+	 *
+	 * @const  string
+	 * @since  __DEPLOY_VERSION__
+	 */
+	const PHP_HEADING_PROTECTION = '<?php die("Access Denied"); ?>';
+
+	/**
+	 * Size of the protection heading PHP_HEADING_PROTECTION
+	 * Just avoid calculating the lenght of the protection heading over and over at runtime
```
Comment (Quy) [inline libraries/joomla/cache/storage/file.php]: Typo `lenght`.

File: libraries/joomla/cache/storage/file.php [hunk 2]
```diff
@@ -346,52 +290,12 @@ public static function isSupported()
...
-		if ($data_lock === true)
-		{
-			// Remember resource, flock release lock if you unset/close resource
-			$this->_locked_files[$path] = $_fileopen;
-		}
-
-		$returning->locked = $data_lock;
-
-		return $returning;
+		// Simulates s successful lock, but actually do nothing,
```
Comment (Quy) [inline libraries/joomla/cache/storage/file.php]: Change `s` to `a`.

File: libraries/joomla/cache/storage/file.php [hunk 3]
```diff
@@ -406,26 +310,14 @@ public function lock($id, $group, $locktime)
...
-		if (isset($this->_locked_files[$path]))
-		{
-			$ret = (bool) @flock($this->_locked_files[$path], LOCK_UN);
-			@fclose($this->_locked_files[$path]);
-			unset($this->_locked_files[$path]);
-
-			return $ret;
-		}
-
+		// Simulates s successful unlock, but actually do nothing,
```
Comment (Quy) [inline libraries/joomla/cache/storage/file.php]: Change `s` to `a`.

**Comment by brianteeman** [comment]
@demis-palma please look at the travis errors

File: libraries/joomla/cache/storage/file.php (hunk 1 above)
Comment (demis-palma) [inline libraries/joomla/cache/storage/file.php]: All corrected, thanks.

**Comment by demis-palma** [comment]
@brianteeman sure.

File: tests/unit/suites/libraries/joomla/cache/storage/JCacheStorageFileTest.php [hunk 4]
```diff
@@ -54,4 +54,40 @@ public function testCacheTimeout()
...
+		$_getFilePath = $reflector->getMethod('_getFilePath');
+		$_getFilePath->setAccessible(true);
+		$file_path = $_getFilePath->invoke($this->handler, $this->id, $this->group);
+		// Prepare arbitrary test data
+		$testData = 'testData';
+
+		// Craft a cache buffer file, and ask the cache object to read the data contained in it
+		file_put_contents($file_path, JCacheStorageFile::PHP_HEADING_PROTECTION . $testData . JCacheStorageFile::INTEGRITY_DIGIT);
+		$readback = $this->handler->get($this->id, $this->group);
+		// Decode the data and comare it with the original
```
Comment (Quy) [inline tests/unit/suites/libraries/joomla/cache/storage/JCacheStorageFileTest.php]: Typo `comare`.

**Comment by csthomas** [comment]
In general it is OK.

Changes in lock/unlock methods may have some B/C issue but I do not protest.
Option from cache `locking` is useless now but is should not be any problem. It should be deprecated.

If this changes can go then other storages can be done in the same way and controllers/adapters can forget about `lock/unlock` cache storages/handlers in `get/store` methods.

File: libraries/joomla/cache/storage/file.php [hunk 5]
```diff
@@ -102,39 +98,19 @@ public function contains($id, $group)
...
 		$path  = $this->_getFilePath($id, $group);
-		$close = false;
 
 		if ($checkTime == false || ($checkTime == true && $this->_checkExpire($id, $group) === true))
 		{
-			if (file_exists($path))
+			$data = file_get_contents($path, false, null, self::PHP_HEADING_SIZE);
+			// is_string($data) also implies $data !== false so avoid the latter check
+			// Explicitly check the case of zero length data, which would cause an illegal array element access below [-1]
+			if (is_string($data) && !empty($data))
```
Comment (csthomas) [inline libraries/joomla/cache/storage/file.php]: `if (data)` would be enough.

File: libraries/joomla/cache/storage/file.php [hunk 6]
```diff
@@ -102,39 +98,19 @@ public function contains($id, $group)
...
-						@fclose($_fileopen);
-					}
-
-					if ($data !== false)
-					{
-						// Remove the initial die() statement
-						return str_replace('<?php die("Access Denied"); ?>#x#', '', $data);
-					}
+					// Remove the integrity digit while returning the buffer
+					return rtrim($data, self::INTEGRITY_DIGIT);
```
Comment (csthomas) [inline libraries/joomla/cache/storage/file.php]: Why not use `substr(data, 0, -1)`?

File: tests/unit/suites/libraries/joomla/cache/storage/JCacheStorageFileTest.php [hunk 7]
```diff
@@ -54,4 +54,40 @@ public function testCacheTimeout()
...
+		// to the protected method JCacheStorageFile::_getFilePath()
+		$reflector = new ReflectionObject($this->handler);
+		$_getFilePath = $reflector->getMethod('_getFilePath');
+		$_getFilePath->setAccessible(true);
+		$file_path = $_getFilePath->invoke($this->handler, $this->id, $this->group);
+		// Prepare arbitrary test data
+		$testData = 'testData';
+
+		// Craft a cache buffer file, and ask the cache object to read the data contained in it
+		file_put_contents($file_path, JCacheStorageFile::PHP_HEADING_PROTECTION . $testData . JCacheStorageFile::INTEGRITY_DIGIT);
```
Comment (csthomas) [inline tests/unit/suites/libraries/joomla/cache/storage/JCacheStorageFileTest.php]: I suggest to use `$this->handler->store($this->id, $this->group, $testData);` to test whether `store` works.

**Comment by tampe125** [comment]
I think this is the correct approach to the cache issue. We lock before writing and there is an extensive use of defensive coding.
:+1:

File: libraries/joomla/cache/storage/file.php (hunk 6 above)
Comment (tampe125) [inline libraries/joomla/cache/storage/file.php]: Because it's a completely different function with a completely different behaviour

Comment (csthomas) [inline libraries/joomla/cache/storage/file.php]: @tampe125 I do not understand your approach. If joomla adds 1 char suffix then it should remove 1 char. What is the issue?
If cached data will contain '\0' at the end of data then `rtrim` could remove too much.

File: libraries/joomla/cache/storage/file.php [hunk 8]
```diff
@@ -346,52 +294,12 @@ public static function isSupported()
...
-		$returning->locked = $data_lock;
-
-		return $returning;
+		// Simulates a successful lock, but actually do nothing,
+		// this cache storage engines delegates file locking to a lower level (LOCK_EX).
+		return (object) array(
+			'locked' => true,
+			'locklooped' => false
+		);
 	}
```
Comment (klas) [inline libraries/joomla/cache/storage/file.php]: This does not seem correct, lock/unlock are there for a reason.

File: libraries/joomla/cache/storage/file.php [hunk 9]
```diff
@@ -406,26 +314,14 @@ public function lock($id, $group, $locktime)
...
-			$ret = (bool) @flock($this->_locked_files[$path], LOCK_UN);
-			@fclose($this->_locked_files[$path]);
-			unset($this->_locked_files[$path]);
-
-			return $ret;
-		}
-
+		// Simulates a successful unlock, but actually do nothing,
+		// this cache storage engines delegates file locking to a lower level (LOCK_EX).
 		return true;
```
Comment (klas) [inline libraries/joomla/cache/storage/file.php]: This does not seem correct, lock/unlock are there for a reason and true should be returned only if unlock was sucesfull. Returned value should depend on real state.

**Review by klas — CHANGES_REQUESTED** [review]

**Comment by HLeithner** [comment]
Similar to #15707 

This PR is outdated and have to be updated to fit in to j3. J4 uses a legacy caching system which should be replaced with a 3rd party library compatible PSR-6 / PSR-16 package see https://github.com/joomla-framework/cache

If you are still interested to bring this to joomla we would be happy if you help us working on the new cache layer or if you like you can bring this PR in Sync for 3.10 but since the focus is J4 it would be better to help use with the caching layer.

thanks I'm closing this for now, feel free to reopen.

**Pull request closed without merging**